File: dataflow/pubsub_to_bq.py
import apache_beam as beam
import apache_beam.transforms.window as window
from apache_beam.options.pipeline_options import PipelineOptions
from apache_beam.options.pipeline_options import StandardOptions
from apache_beam.io.gcp.pubsub import ReadFromPubSub
from apache_beam.io.gcp.bigquery import WriteToBigQuery

# Set the project ID, topic and subscription name
project = 'my-project'
topic_name = 'my-topic'
subscription_name = 'my-subscription'

dataset_id = 'my-dataset'
table_id = 'my-table'
table_schema = 'field1:STRING,field2:INTEGER,field3:FLOAT'

options = PipelineOptions()
options.view_as(StandardOptions).runner = 'DataflowRunner'

with beam.Pipeline(options=options) as pipeline:

    messages = (pipeline
                | 'ReadFromPubSub' >> ReadFromPubSub(subscription=f'projects/{project}/subscriptions/{subscription_name}')
                | 'DecodeMessage' >> beam.Map(lambda message: message.decode('utf-8')))

    (messages
     | 'ParseJSON' >> beam.Map(lambda message: eval(message))
     | 'AddTimestamp' >> beam.Map(lambda element: beam.window.TimestampedValue(element, element['timestamp']))
     | 'Window' >> beam.WindowInto(window.FixedWindows(1))
     | 'WriteToBigQuery' >> WriteToBigQuery(table=table_id, dataset=dataset_id, schema=table_schema))

File: dataflow/pubsub_to_bq.java
import com.google.cloud.dataflow.sdk.Pipeline;
import com.google.cloud.dataflow.sdk.io.gcp.bigquery.BigQueryIO;
import com.google.cloud.dataflow.sdk.io.gcp.bigquery.TableSchema;
import com.google.cloud.dataflow.sdk.io.gcp.pubsub.PubsubIO;
import com.google.cloud.dataflow.sdk.options.Default;
import com.google.cloud.dataflow.sdk.options.Description;
import com.google.cloud.dataflow.sdk.options.PipelineOptions;
import com.google.cloud.dataflow.sdk.options.PipelineOptionsFactory;
import com.google.cloud.dataflow.sdk.transforms.DoFn;
import com.google.cloud.dataflow.sdk.transforms.MapElements;
import com.google.cloud.dataflow.sdk.transforms.ParDo;
import com.google.cloud.dataflow.sdk.transforms.windowing.AfterWatermark;
import com.google.cloud.dataflow.sdk.transforms.windowing.FixedWindows;
import com.google.cloud.dataflow.sdk.transforms.windowing.Window;
import com.google.cloud.dataflow.sdk.values.PCollection;
import com.google.cloud.dataflow.sdk.values.PDone;
import com.google.common.collect.ImmutableList;

@Description("The Cloud Pub/Sub topic to read from.")
@Required
String getInputTopic();

void setInputTopic(String value);

@Description("The BigQuery table to write the results to.")
@Required
String getOutputTable();

void setOutputTable(String value);

@Description("The BigQuery table to write invalid data to.")

public class TumblingWindowAveragePipeline {

public static void main(String[] args) throws Exception {
PipelineOptionsFactory.register(TumblingWindowAveragePipelineOptions.class);
TumblingWindowAveragePipelineOptions options = PipelineOptionsFactory.fromArgs(args).as(TumblingWindowAveragePipelineOptions.class);
  
Pipeline pipeline = Pipeline.create(options);

PCollection<String> messages = pipeline
    .apply(PubsubIO.readFromPubSub(options.getInputTopic()));

PCollection<Double> averages = messages
    .apply(ParDo.of(new DoFn<String, Double>() {
      @ProcessElement
      public void processElement(ProcessContext c) {
        String message = c.element();
        String[] parts = message.split(",");
        double average = Double.parseDouble(parts[1]) / Double.parseDouble(parts[0]);
        c.output(average);
      }
    }))
    .apply(Window.into(FixedWindows.of(Duration.standardMinutes(1))));

averages.apply(BigQueryIO.writeTableRows(options.getOutputTable(), new TableSchema()
    .setFields(ImmutableList.of(
        new TableSchema.Field("average", TableSchema.FieldType.FLOAT64)
    ))));

averages.apply(MapElements.via(new DoFn<Double, Double>() {
  @ProcessElement
  public void processElement(ProcessContext c) {
    Double average = c.element();
    if (average < 0) {
      c.output(null);
    } else {
      c.output(average);
    }
  }
})).apply(AfterWatermark.pastEndOfWindow())
    .apply(BigQueryIO.writeTableRows(options.getDeadletterTable(), new TableSchema()
        .setFields(ImmutableList.of(
            new TableSchema.Field("average", TableSchema.FieldType.FLOAT64)
        ))));

pipeline.run();
  
  
